### ISSUE TITLE
Noisy ECAL DQM in gpu validation workflows

### ISSUE THREAD

**Issue by mmusich** [cms-sw/cmssw#42720]
When adding a new workflow for patatrack validation on 2023 data (see PR https://github.com/cms-sw/cmssw/pull/42674) the ECAL DQM for the gpu task is very noisy, emitting several times per event this sort of warnings:

```console
Begin processing the 1st record. Run 366727, Event 132255498, LumiSection 89 on stream 0 at 30-Aug-2023 11:19:08.112 CEST
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
EcalRawDataCollection does not exist. No event-type filtering will be applied
%MSG
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
Ecal Monitor Source::runOnCollection: EcalRawData does not exist
%MSG
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
Ecal Monitor Source::runOnCollection: TrigPrimEmulDigi does not exist
%MSG
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
Ecal Monitor Source::runOnCollection: EBGpuRecHit does not exist
%MSG
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
Ecal Monitor Source::runOnCollection: EEGpuRecHit does not exist
%MSG
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
Ecal Monitor Source::runOnCollection: EBReducedRecHit does not exist
%MSG
%MSG-w EcalDQM:  EcalDQMonitorTask:ecalMonitorTaskEcalOnly  30-Aug-2023 11:19:10 CEST Run: 366727 Event: 132255498
Ecal Monitor Source::runOnCollection: EEReducedRecHit does not exist
%MSG
```

These appear also in the `DQM/Integration` unit tests for the ECAL GPU client (introduced in PR https://github.com/cms-sw/cmssw/pull/42542 ) which is run in IBS:

see e.g. a log for [CMSSW_13_3_X_2023-09-04-1100](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc11/CMSSW_13_3_X_2023-09-04-1100/unitTestLogs/DQM/Integration#/6257)

Could core DQM / ECAL DQM experts have a look?

**Comment by mmusich** [comment]
assing dqm, ecal-dpg

**Comment by alejands** [comment]
@mmusich Do we expect the ECAL GPU validation plots to be filled for these unit tests? If not, we can we suppress the number of warnings to avoid spam

**Comment by mmusich** [comment]
@alejands 

>  Do we expect the ECAL GPU validation plots to be filled for these validation tests? 

Yes, we do.  That's the whole purpose of the validation workflow.

**Comment by makortel** [comment]
assign heterogeneous

**Comment by abhih1** [comment]
> @alejands
> 
> > Do we expect the ECAL GPU validation plots to be filled for these validation tests?
> 
> Yes, we do. That's the whole purpose of the validation workflow.

@mmusich I think what we wanted to clarify was whether the GPU validation plots are indeed getting filled for these tests despite these warnings.
We suspect these warnings are coming from the difference in the inputsource for the unit tests, a change recently made through this PR: https://github.com/cms-sw/cmssw/commit/56c6da4344da99a533cb3ed7149fbba0cd76aa11
 
 If the plots are getting filled, then we could suppress the number of warnings to reduce the noise.

**Comment by mmusich** [comment]
> We suspect these warnings are coming from the difference in the inputsource for the unit tests, a change recently made through this commit:

That test uses DQM streamer files as input, which is exactly what the online DQM sees at P5. So the input collections should be adapted in order to compute with that.  Also what about the relval validation workflow that uses regular repacked RAW data introduced in #42674?
That cannot depend on commit 56c6da4.

> was whether the GPU validation plots are indeed getting filled for these tests despite these warnings.

I can't really comment on that.  I would invite you to run one of these tests locally and check. Hope this helps.

**Comment by alejands** [comment]
@mmusich I was able to fix a bug causing the warnings for the ECAL GPU validation task and verified it using the tests introduced in https://github.com/cms-sw/cmssw/pull/42542, but I was unable to reproduce the error in https://github.com/cms-sw/cmssw/pull/42674 on the lxplus-gpu node. 

I keep getting errors like this when trying to use CMSSW on the lxplus-gpu node:
```
/cvmfs/cms.cern.ch/slc7_amd64_gcc11/external/git/2.38.1-ee97e960a104e95b9f0c52a98b85ce43/libexec/git-core/git-remote-https: error while loading shared libraries: libssl.so.10: cannot open shared object file: No such file or directory
```

Some of the warnings for https://github.com/cms-sw/cmssw/pull/42674 don't seem to be related to the GPU validation module, so I would like to reproduce the error myself in order to find the culprit.

**Comment by mmusich** [comment]
> I keep getting errors like this when trying to use CMSSW on the lxplus-gpu node:

When developing for https://github.com/cms-sw/cmssw/pull/42674 I was using an `lxplus8-gpu` node, though I don't immediately see why the arch should matter. Maybe you can try to test there.

**Comment by fwyzard** [comment]
I just noticed that some (all ?) `lxplus-gpu` nodes are running Red Hat Enterprise Linux release 9.2 -- maybe that's why the SLC7 version of `git` is having troubles.

**Comment by alejands** [comment]
> When developing for #42674 I was using an `lxplus8-gpu` node, though I don't immediately see why the arch should matter. Maybe you can try to test there.

I was able to run wf `141.008583` on the `lxplus-gpu` and debug, thanks!

**Comment by alejands** [comment]
For wf `141.008583`, there were two somewhat separate but related issues related to which collections were trying to be used. Some collections are modified for online GPU validation, but not for offline.

1. The first issue has to do with looking for ECAL GPU rec hits. As far as I know, these are not yet implemented, so naturally it won't find any GPU rec hit collections. A placeholder input tag is in place, but I believe the GPU tags in particular don't actually point to anything. Perhaps @thomreis can comment on this. 

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/EcalMonitorTasks/python/CollectionTags_cfi.py#L56-L59

Rec hit GPU validation was turned on in the GPU module by default, but it is manually turned off in the Online ECAL DQM GPU client `ecalgpu_dqm_sourceclient-live_cfg.py`. The online client is also set to only run the GPU validation module and no other ECAL DQM modules.

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/Integration/python/clients/ecalgpu_dqm_sourceclient-live_cfg.py#L69-L71

2. If we only care to run the GPU validation module here as well, then that should also solve *most* of the second issue. In online GPU validation, there are modified input tags for some collections, but again there is no such modification done for offline validation. 

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/Integration/python/clients/ecalgpu_dqm_sourceclient-live_cfg.py#L77-L86

If we are only running the GPU validation module (except for rec hits), then the only input tag complaining is for `EcalRawData`. 

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/EcalMonitorTasks/python/CollectionTags_cfi.py#L12

This collection is unused by the GPU module, but this input tag is always checked to run any ECAL DQM processing for historical reasons. The ECAL digis for GPU validation use the same collection, but the input tags include the `@cpu` and `@cuda` modifiers. 

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/EcalMonitorTasks/python/CollectionTags_cfi.py#L48-L51

These tags are being used and are not complaining, so would it be okay to customize `EcalRawData` to `ecalDigis@cpu` here for GPU validation? I want to make sure there's no nuance to worry about here.

2. (b) If we *do* care about running other modules, then I believe the issue also has to do with a difference in input tags. The remaining collection tags that are complaining are

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/EcalMonitorTasks/python/CollectionTags_cfi.py#L31

https://github.com/cms-sw/cmssw/blob/bac24a49aba966c5c09262c4e9383473266cd2b6/DQM/EcalMonitorTasks/python/CollectionTags_cfi.py#L42-L43

If they are being used here, are these the correct input tags? If not, then we don't need to worry about these. The warning should go away by only enabling the GPU module.

**Comment by fwyzard** [comment]
hi @alejands,
I think we should be able to run the full ECAL DQM on this workflows.

Does it work for the ECAL only configurations (.511 for MC) ?

**Comment by mmusich** [comment]
> Does it work for the ECAL only configurations (.511 for MC) ?

as far as I can tell, similar warnings are visible in .511 workflows in IBs too: https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el9_amd64_gcc11/CMSSW_13_3_X_2023-09-06-2300/pyRelValMatrixLogs/run/11634.511_TTbar_14TeV+2021_Patatrack_ECALOnlyCPU/step3_TTbar_14TeV+2021_Patatrack_ECALOnlyCPU.log#/

**Comment by alejands** [comment]
I was able to modify one of the cfg files produced by WF `141.008583` and get all the ECAL collections being produced during the RECO step.

For the ECAL Trigger Primitives, there are no emulated digi collections available, only the standard ones.
```
edm::SortedCollection<EcalTriggerPrimitiveDigi,edm::StrictWeakOrdering<EcalTriggerPrimitiveDigi> >    "ecalDigis"                 "EcalTriggerPrimitives"   "reRECO"
```

There are also no Reduced Ecal Rec Hits available. Fortunately for both of these cases, we can easily toggle these off and eliminate the warnings.

https://github.com/cms-sw/cmssw/blob/a4f12bbcb405b038a0a2bca889d88eb87efe5f3e/DQM/EcalMonitorTasks/python/TrigPrimTask_cfi.py#L33

https://github.com/cms-sw/cmssw/blob/a4f12bbcb405b038a0a2bca889d88eb87efe5f3e/DQM/EcalMonitorTasks/python/RecoSummaryTask_cfi.py#L7

However, for the case of `EcalRawData`, we require an `EcalRawDataCollection` (wrapper for an `EcalDCCHeaderBlock` collection) which is not available. They are, however, available for the ECAL Preshower. 
```
edm::SortedCollection<ESDCCHeaderBlock,edm::StrictWeakOrdering<ESDCCHeaderBlock> >    "ecalPreshowerDigis"        ""                "reRECO"
```
It appears that this collection is more important for several ECAL DQM modules, but as far as I know, ECAL DQM is not directly in charge of which collections are produced.

**Comment by mmusich** [comment]
> There is the question of at what level do we want to change these flags? (eg. in the ECAL DQM `cfi` files in `DQM/EcalMonitorTasks`, at the level of the cfi and cff files in `DQMOffline/Ecal`, or at the WF level)

I guess this depends if these flags are useful in other setups (and this only ECAL DQM knows about)

> It appears that this collection is more important for several ECAL DQM modules, but as far as I know, ECAL DQM is not directly in charge of which collections are produced.

maybe (but ECAL DPG at large certainly is).

**Comment by thomreis** [comment]
I have updated the ECAL unpacker CPUDigis module to produce dummy collections for the `EcalRawDataCollection` and some other collections that the CPU unpacker can produce but the GPU one could not. See PR #42844.
This should fix the `EcalRawData` warning messages.

**Comment by alejands** [comment]
I implemented the changes discussed earlier in https://github.com/cms-sw/cmssw/pull/42848. Before knowing about the fix for `EcalRawData` in https://github.com/cms-sw/cmssw/pull/42844, I implemented a customization option to skip collections and not consume them on the ECAL DQM end. While this is no longer needed, we decided to leave it in as it could prove useful for something else in the future, though it's not being used right now.

**Comment by mmusich** [comment]
> I believe this issue is now resolved and can be closed.

It is lacking signatures from the involved groups

**Comment by thomreis** [comment]
+ecal-dpg

**Comment by mmusich** [comment]
Solved in #42844 + https://github.com/cms-sw/cmssw/pull/42848
